xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,785 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
 func (bb) Name() string {
 	return Name
 }
+
+// scUpdate wraps a subConn update to be sent to the child balancer.
+type scUpdate struct {
+	scw   *subConnWrapper
+	state balancer.SubConnState
+}
+
+type ejectedUpdate struct {
+	scw     *subConnWrapper
+	ejected bool // true for ejected, false for unejected
+}
+
+type outlierDetectionBalancer struct {
+	numAddrsEjected int // For fast calculations of percentage of addrs ejected
+
+	childState       balancer.State
+	recentPickerNoop bool
+
+	closed *grpcsync.Event
+	cc     balancer.ClientConn
+	bOpts  balancer.BuildOptions
+
+	// childMu protects child and also updates to the child (to uphold the
+	// balancer.Balancer API guarantee of synchronous calls). It also protects
+	// against run() reading that the child is not nil for SubConn updates, and
+	// then UpdateClientConnState or Close writing to the the child.
+	childMu sync.Mutex
+	child   balancer.Balancer
+
+	// mu guards access to a lot of the core LB Policy State. It also prevents
+	// intersplicing certain operations.
+	//
+	// ex 1: interval timer goes off, outlier detection algorithm starts running
+	// based on knobs in cfg. in the middle of running the algorithm, a
+	// ClientConn update comes in and writes to cfg. This causes undefined
+	// behavior for the interval timer algorithm.
+	//
+	// ex 2: Updating the addrs map from UpdateAddresses in the middle of
+	// running the interval timer algorithm which uses addrs heavily. This will
+	// cause undefined behavior for the interval timer algorithm.
+	mu             sync.Mutex
+	addrs          map[string]*addressInfo
+	cfg            *LBConfig
+	scWrappers     map[balancer.SubConn]*subConnWrapper
+	timerStartTime time.Time
+	intervalTimer  *time.Timer
+
+	scUpdateCh     *buffer.Unbounded
+	pickerUpdateCh *buffer.Unbounded
+}
+
+// noopConfig returns whether this balancer is configured with a logical no-op
+// configuration or not.
+func (b *outlierDetectionBalancer) noopConfig() bool {
+	return b.cfg.SuccessRateEjection == nil && b.cfg.FailurePercentageEjection == nil
+}
+
+func (b *outlierDetectionBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*LBConfig)
+	if !ok {
+		return balancer.ErrBadResolverState
+	}
+
+	// Reject whole config if any errors, don't persist it for later
+	bb := balancer.Get(lbCfg.ChildPolicy.Name)
+	if bb == nil {
+		return fmt.Errorf("outlier detection: child balancer %q not registered", lbCfg.ChildPolicy.Name)
+	}
+
+	if b.child == nil || b.cfg.ChildPolicy.Name != lbCfg.ChildPolicy.Name {
+		b.childMu.Lock()
```

Comment:
Unless I'm mistaken and there's some other implied synchronization going on, you need to be holding `b.childMu` _any_ time you access `b.child`.  In much of this file, you read it before taking the lock.

Reply:
Oh, and also I just remembered, this is documented but the reason I grab it before forwarding updates is the balancer.Balancer API has a guarantee of no concurrent calls, and holding this mutex before calling downward into child makes it so only one operation on the child is called at a time.

Reply:
Can this moved down to the place where configs are pushed to the child policies? That way everything about the child policies in this function can be done with the lock held. And as an added benefit, they will appear as one block of code, easing cognitive load.

Reply:
Thanks for the suggestion. However, I tried this and it doesn't work. If this is the first update, b.cfg.ChildPolicy will throw a panic (cfg will be nil, and ChildPolicy is also nil even on a default cfg). However, with my b.child == nil check, the first conditional will hit True on the first update and not check the second conditional.

Reply:
Why is it OK to access `b.cfg` here without holding `b.mu`?

Reply:
> Right now, the interval timer algorithm (think interval being zero)/uneject all subconns in the case of flipping to a no-op config is guaranteed to read this new b.child as it is forwarding updates downward (b.UpdateSubConnState)

With the current state of the code, the following can happen:
- we receive the config, and close the old child policy and build the new one
- update the address map, and as part of that figure out that we have a no-op config, and hence uneject all previously ejected subConns
  - these will show up as subConnState updates on the new child

But we still haven't given the child it's configuration. We do that only in line 288. This doesn't seem right to me.

The gRFC states the following:
```
When the outlier_detection LB policy receives an address update, it will create a map entry
for each subchannel address in the list, and remove each map entry for a subchannel address
not in the list, then pass the address list along to the child policy.
```

The correct order then, it seems would be:
- update the address maps in the balancer based on the received update
- close the old child and create a new one if the child policy name has changed
- push the new config to the child
- do the stuff around stopping/starting the interval timer algorithm, unejecting addresses etc